processing/src/main/java/io/druid/query/extraction/LookupReferencesManager.java

```diff
@@ -0,0 +1,169 @@
+  {
+    synchronized (lock) {
+      assertStarted();
+      if (!lookupExtractorFactory.start()) {
+        throw new ISE("start method returned false for lookup [%s]", lookupName);
+      }
+      return (null == lookupMap.putIfAbsent(lookupName, lookupExtractorFactory));
+    }
+  }
+
+  /**
+   * @param lookups {@link ImmutableMap<String, LookupExtractorFactory>} containing all the lookup as one batch.
+   *
+   * @throws ISE if the manager is closed or if {@link LookupExtractorFactory#start()} returns false
+   */
+  public void put(ImmutableMap<String, LookupExtractorFactory> lookups) throws ISE
+  {
+    synchronized (lock) {
+      assertStarted();
+      for (ImmutableMap.Entry<String, LookupExtractorFactory> entry : lookups.entrySet()) {
+        final String lookupName = entry.getKey();
+        final LookupExtractorFactory lookupExtractorFactory = entry.getValue();
+        if (!lookupExtractorFactory.start()) {
+          throw new ISE("start method returned false for lookup [%s]", lookupName);
+        }
+        if (null != lookupMap.putIfAbsent(lookupName, lookupExtractorFactory)) {
+          LOGGER.warn("lookup with name [%s] is not add since it already exist", lookupName);
+        }
+      }
+    }
+  }
+
+  /**
+   * @param lookupName namespace of {@link LookupExtractor} to delete from the reference registry.
+   *                   this function does call the cleaning method {@link LookupExtractor#close()}
+   *
+   * @return true only if {@code lookupName} is removed and the lookup correctly stopped
+   */
+  public boolean remove(String lookupName)
+  {
+    final LookupExtractorFactory lookupExtractorFactory = lookupMap.remove(lookupName);
+    if (lookupExtractorFactory != null) {
+      LOGGER.debug("Removing lookup [%s]", lookupName);
+      return lookupExtractorFactory.stop();
+    }
+    return false;
+  }
+
+  /**
+   * @param lookupName namespace key to fetch the reference of the object {@link LookupExtractor}
+   *
+   * @return reference of {@link LookupExtractor} that correspond the the namespace {@code lookupName} or null if absent
+   *
+   * @throws ISE if the {@link LookupReferencesManager} is closed or did not start yet
+   */
+  @Nullable
+  public LookupExtractorFactory get(String lookupName) throws ISE
+  {
+    final LookupExtractorFactory lookupExtractorFactory = lookupMap.get(lookupName);
+    assertStarted();
+    return lookupExtractorFactory;
+  }
+
+  /**
+   * @return Returns {@link ImmutableMap} containing a copy of the current state.
+   *
+   * @throws ISE if the is is closed or did not start yet.
+   */
+  public ImmutableMap<String, LookupExtractorFactory> getAll() throws ISE
+  {
+    assertStarted();
+    return ImmutableMap.copyOf(lookupMap);
+  }
+
+  private void assertStarted() throws ISE
+  {
+    if (isClosed()) {
+      throw new ISE("lookup manager is closed");
+    }
+  }
+
+  @VisibleForTesting
```

Comment:
I'm generally not a fan of "visible for testing" stuff.  Couldn't this just be public?